runtime/src/main/cpp/Memory.cpp

```diff
@@ -2252,6 +2253,56 @@ void FreezeSubgraph(ObjHeader* root) {
 #endif
 }
 
+void deepFrozenCopyTo(ObjHeader* to, ObjHeader* from, ObjHeaderMap* translationMap);
+
+void changeToCopyIfNeeded(ObjHeader** location, ObjHeaderMap* translationMap) {
+  auto* object = *location;
+  if (object == nullptr) return;
+  if (PermanentOrFrozen(object)) {
+    AddHeapRef(object);
+    return;
+  }
+  auto substitution = translationMap->find(object);
+  if (substitution != translationMap->end()) {
+    *location = substitution->second;
+    AddHeapRef(*location);
+  } else {
+    ObjHolder result;
+    auto* clone = AllocInstance(object->type_info(), result.slot());
```

Comment:
IIRC, we are considering making this part opt-in, i.e. enabled only for certain classes.
Do you plan to change this later thus breaking the behaviour if we decide in favor of opt-in?

Reply:
Yeah, we just haven't decided how to distinguish such types, so idea here is to try how fitting the solution is.